url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/syntax/builder/SyntaxBuilderSemantic.java

```diff
@@ -0,0 +1,195 @@
+	The leading must be the one with the lower UNICODE value.
+	<p>
+	The semantics of "-" is intersection. When specifiying <i>chars - comment - stringdef</i> this means
+	all <i>chars</i> but not <i>comments</i> or <i>stringdef</i>, i.e. <i>stringdef</i> is not subtracted
+	from <i>comment</i> but from <i>chars</i>!
+	<p>
+	TODO: think over repeat number symbol: parser AND lexer would need this.
+			Better define this by written symbols like "a ::= b b b b;"? But what to do on hundred "b"?
+	
+	@author (c) 2002 Fritz Ritzberger
+*/
+
+public class SyntaxBuilderSemantic implements Semantic
+{
+	public static String [][] syntax = {
+		// START rule
+		{ "syntax",    "syntax", "rule" },	// the grammar consists of rules
+		{ "syntax",    "rule" },
+
+		{ "set",     "`bnf_chardef`", "\""+Token.UPTO+"\"", "`bnf_chardef`" },	// set of characters
+
+		{ "intersectionstartunit",  "set" },	// intersection of character sets
+		{ "intersectionstartunit",  "`identifier`" },
+		{ "intersectionstartunit",  "`ruleref`" },
+
+		{ "intersectionunit",  "`bnf_chardef`" },
+		{ "intersectionunit",  "`stringdef`" },
+		{ "intersectionunit",  "intersectionstartunit" },
+
+		{ "intersectionsubtracts", "intersectionsubtracts", "intersectionsubtract" },
+		{ "intersectionsubtracts", "intersectionsubtract" },
+		{ "intersectionsubtract", "'"+Token.BUTNOT+"'", "intersectionunit" },
+
+		{ "intersection", "intersectionstartunit", "intersectionsubtracts" },
+
+		{ "sequnit",  "intersection" },	// unit of a sequence
+		{ "sequnit",  "intersectionunit" },
+		{ "sequnit",  "'('", "unionseq", "')'" },
+
+		{ "quantifiedsequnit", "sequnit", "`quantifier`" },	// unit can be quantified
+		{ "quantifiedsequnit", "sequnit" },
+
+		{ "sequence", "sequence", "quantifiedsequnit" },	// sequence of units with significant order
+		{ "sequence", "quantifiedsequnit" },
+
+		{ "sequence_opt", "sequence" },	// sequence is nullable
+		{ "sequence_opt", /*nothing*/ },
+
+		{ "unionseq", "unionseq", "'|'", "sequence_opt" },	// rule alternatives
+		{ "unionseq", "sequence_opt" },
+
+		{ "rule",    "`identifier`", "\"::=\"", "unionseq", "';'" },	// one rule of a grammar
+
+		// specify what will be ignored. Using StandardLexerRules.lexerSyntax will include this automatically
+		//{ "ignored", "`comment`" },
+		//{ "ignored", "`spaces`" },
+		//{ "ignored", "`newlines`" },
+	};
+	
+	private List initialNonterminals;
+	
+
+	/**
+		Creates a syntax builder semantic that resolves parenthesis and quantifiers.
+	*/
+	public SyntaxBuilderSemantic()	{
+		this(null);
+	}
+	
+	/**
+		Creates a syntax builder semantic that resolves parenthesis and quantifiers.
+		All nonterminals read from the processed syntax will be collected into the passed List.
+	*/
+	public SyntaxBuilderSemantic(List initialNonterminals)	{
+		this.initialNonterminals = initialNonterminals;
+	}
+	
+
+	public Object doSemantic(Rule rule, List inputTokens, List ranges)	{
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.